src/home/homework.c

```diff
@@ -1795,8 +1798,31 @@ static int home_inspect(UserRecord *h, UserRecord **ret_home) {
         return 1;
 }
 
+static int user_session_freezer(uid_t uid, bool freeze, UnitFreezer *ret) {
+        _cleanup_free_ char *unit = NULL;
+
+        // TODO: Figure out if we should be freezing:
+        //   user@<UID>.service, OR
+        //   user-<UID>.slice
+        //
+        // Freezing the slice breaks GDM re-authentication, because it
+        // contains session-<ID>.scope, which contains gdm-session-worker.
+        // This is used by GDM to re-authenticate and unlock a session. Thus,
+        // if we freeze the whole user-<UID>.slice, there's no reauth worker.
+        // HOWEVER, I think conceptually it makes more sense to freeze the whole
+        // slice. Will need to discuss w/ GDM maintainers
+        if (asprintf(&unit, "user@" UID_FMT ".service", uid) < 0)
+                return log_oom();
+
+        if (freeze)
+                return unit_freezer_freeze(unit, ret);
+        else
+                return unit_freezer_recover(unit, ret);
+}
```

Comment:
hmm, why do we need this at all btw? i mean, the sleep logic already freezes user.slice, hence why bother with freezing the per-user slices around homectl lock?

I mean, the purpose of the home dir locking is really just the sleep logic, hence i see no reason why we would treat this as a separate concept, and also freeze the per-user slice here.

Reply:
Because `homectl lock` can [and will](https://gitlab.gnome.org/Teams/STF/homed/-/issues/4) happen outside of suspend

Ultimately I'd also like to rework systemd-sleep to _not_ call `LockAllHomes`, and make the desktop environment responsible for deciding when to lock itself. In other words I want do decouple locking from sleep.

Reply:
We've discussed this over a call.

The solution we settled on is that the desktop environment should be able to call a special method that hands over an inhibitor FD and inhibits `LockAllHomes` for that user. This way the desktop environment can opt-out of `LockAllHomes` if it wants to do something fancy. However if something goes wrong in the DE then it can close the FD (which will happen implicitly if it crashes too!) and homed's `LockAllHomes` behavior will fail closed.

I imagine that GNOME specifically will have a special `gnome-settings-daemon` service that will manage the fancy locking behavior, which is great because then if something goes wrong it can just exit and security of the session will be preserved